File: GeoNoter/Models/PQGPoint.m
//
//  PQGPoint.m
//  GeoNoter
//
//  Created by Patrick Quinn-Graham on 20/07/2014.
//  Copyright (c) 2014 Patrick Quinn-Graham. All rights reserved.
//

#import <AddressBookUI/AddressBookUI.h>

#import "GeoNoter-Swift.h"
#import "PersistStore.h"

#import "NSDateJSON.h"
#import "NSStringUUID.h"

#import "PQGPoint.h"

@interface PQGPoint()
  @property (nonatomic, copy) NSString *friendly_name;
  @property (nonatomic, copy) NSDate *recorded_at;
@end

@implementation PQGPoint

+ (NSArray*)allInstances {
  return [self instancesOrderedBy:@"name ASC"];
}

#pragma mark - Map underscore to camelcase

- (NSString*)friendlyName {
  return self.friendly_name;
}

- (void)setFriendlyName:(NSString *)friendlyName {
  self.friendly_name = friendlyName;
}

- (NSDate*)recordedAt {
  return self.recorded_at;
}

- (void)setRecordedAt:(NSDate *)recordedAt {
  self.recorded_at = recordedAt;
}


//- (NSString *)description
//{
//  return [NSString stringWithFormat:@"<%@> ID: '%@'. Name: '%@'.", [self class], self.dbId, self.name];
//}

#pragma mark - FCModel

- (id)serializedDatabaseRepresentationOfValue:(id)instanceValue forPropertyNamed:(NSString *)propertyName {
  if([propertyName isEqualToString:@"recorded_at"]) {
    NSDate *instanceDate = instanceValue;
    return [instanceDate pqg_sqlDateString];
  } else {
    return [super serializedDatabaseRepresentationOfValue:instanceValue forPropertyNamed:propertyName];
  }
}

- (id)unserializedRepresentationOfDatabaseValue:(id)databaseValue forPropertyNamed:(NSString *)propertyName {
  if([propertyName isEqualToString:@"recorded_at"]) {
    return [NSDate pqg_dateWithSQLString:databaseValue];
  } else {
    return [super unserializedRepresentationOfDatabaseValue:databaseValue forPropertyNamed:propertyName];
  }
}

#pragma mark - Tags

-(NSArray*)tags {
  return [PQGTag instancesWhere:@"id IN (select tag_id FROM point_tag WHERE point_id = ?) ORDER BY name ASC", self.id];
}

-(void)addTag:(PQGTag*)tag {
//  [FCD]
  DLog(@"Add tag... %lld", tag.id);
  [FCModel executeUpdateQuery:@"INSERT INTO point_tag (tag_id, point_id) VALUES (?, ?)", tag.id, self.id];
}

-(void)removeTag:(PQGTag*)tag {
  DLog(@"Removing exist tag... %lld", tag.id);
  [FCModel executeUpdateQuery:@"DELETE FROM point_tag WHERE tag_id = ? AND point_id = ?", tag.id, self.id];
}

#pragma mark - Attachments

-(NSArray*)attachments {
  return [PQGAttachment instancesWhere:@"point_id = ? ORDER BY recorded_at ASC", self.id];
}

-(PQGAttachment*)addAttachment:(NSData*)attachment withExtension:(NSString *)extension {
  
  NSString *fileName = [[NSString stringWithUUID] stringByAppendingPathExtension:extension];
  NSString *actualFile = [[PersistStore attachmentsDirectory] stringByAppendingPathComponent:fileName];
  
  [attachment writeToFile:actualFile atomically:YES];
  
  PQGAttachment *newAttachment = [PQGAttachment new];

  newAttachment.fileName = fileName;
  newAttachment.kind = extension;
  newAttachment.pointId = self.id;
  
  NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
  dateFormatter.dateStyle = NSDateFormatterMediumStyle;
  dateFormatter.timeStyle = NSDateFormatterMediumStyle;
  
  NSDate *today = [NSDate date];
  
  newAttachment.friendlyName = [@"Picture - " stringByAppendingString:[dateFormatter stringFromDate:today]];
  newAttachment.memo = @"No memo";
  newAttachment.recordedAt = today;
  [newAttachment save];
  
  return newAttachment;
  
}

#pragma mark - Defaults


-(void)determineDefaultName:(NSString*)locationName {
  if ([[[NSUserDefaults standardUserDefaults] stringForKey:@"LocationsDefaultName"] isEqualToString:@"most-specific"] && locationName != nil) {
    NSLog(@"Use location name... %@", locationName);
    self.name = locationName;
  } else if([[[NSUserDefaults standardUserDefaults] stringForKey:@"LocationsDefaultName"] isEqualToString:@"coords"]) {
    self.name = [NSString stringWithFormat:@"%f, %f", self.longitude, self.latitude];
    NSLog(@"Use lat/long name... %@", self.name);
  } else {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterMediumStyle];
    [dateFormatter setTimeStyle:NSDateFormatterMediumStyle];
    self.name = [dateFormatter stringFromDate:[NSDate date]];
    NSLog(@"Use date/time name... %@", self.name);
  }
  
}

#pragma mark - Geocode (maybe)

- (void)setupAsNewItem:(void (^)())completionHandler {
  
  PQGLocationHelper *helper = [PQGLocationHelper sharedHelper];
  
  [helper location:^(CLLocation *location, NSError *error) {
    if(error) {
      NSLog(@"Oh oh. Could not get location. Should explode.");
    } else {
      NSLog(@"long! %f", location.coordinate.longitude);
      NSLog(@"lat! %f", location.coordinate.latitude);
      
      self.latitude = location.coordinate.latitude;
      self.longitude = location.coordinate.longitude;
      self.recordedAt = [NSDate date];
      self.name = @"Untitled";
      self.friendlyName = @"Untitled";
      self.memo = @"No memo";
      
      if([[NSUserDefaults standardUserDefaults] boolForKey:@"LocationsUseGeocoder"]) {
        NSLog(@"Geocoder time!");
        [self geocode:^(NSError *error){
          [self save];
          completionHandler();
        }];
      } else {
        NSLog(@"No geocoder here baby! %@", [[NSUserDefaults standardUserDefaults] boolForKey:@"LocationsUseGeocoder"] ? @"YES" : @"NO");
        [self determineDefaultName:nil];
        [self save];
        completionHandler();
      }
      
    }
  }];
}

#pragma mark -
#pragma mark Geocoder

-(void)geocode:(void (^)(NSError*))completionHandler {
  //  UIApplication *app = [UIApplication sharedApplication];
  //  NSAssert(self->bgTask == UIBackgroundTaskInvalid, nil);
  //
  //  self->bgTask = [app beginBackgroundTaskWithExpirationHandler:^{
  //  }];
  //		if (self->bgTask != UIBackgroundTaskInvalid) {
  //      [[UIApplication sharedApplication] endBackgroundTask:self->bgTask];
  //      self->bgTask = UIBackgroundTaskInvalid;
  //    }
  
  CLLocation *location = [[CLLocation alloc] initWithLatitude:self.latitude longitude:self.longitude];
  
  NSLog(@"Geocoding %@", location);
  
  [[PQGLocationHelper sharedHelper] geocode:location completionHandler:^(NSArray *placemarks, NSError *error) {
    if(error || placemarks.count == 0) {
      self.friendlyName = @"Geocoder Unavailable";
    } else {
      [self reverseGeocoderDidFindPlacemark:placemarks[0]];
    }
    completionHandler(error);
  }];
}

- (void)reverseGeocoderDidFindPlacemark:(CLPlacemark *)placemark {
  NSString *simpleName = placemark.country;
  
  if(placemark.administrativeArea && ![placemark.administrativeArea isEqualToString:@""]) {
    simpleName = placemark.administrativeArea;
  }
  if(placemark.locality && ![placemark.locality isEqualToString:@""]) {
    simpleName = placemark.locality;
  }
  if(placemark.subLocality && ![placemark.subLocality isEqualToString:@""]) {
    simpleName = placemark.subLocality;
  }
  if(placemark.thoroughfare && ![placemark.thoroughfare isEqualToString:@""]) {
    if(!placemark.subLocality || [placemark.subLocality isEqualToString:@""]) {
      simpleName = placemark.thoroughfare;
    }
  }
  if(placemark.subThoroughfare && ![placemark.subThoroughfare isEqualToString:@""]) {
    if(!placemark.subLocality || [placemark.subLocality isEqualToString:@""]) {
      simpleName = [placemark.subThoroughfare stringByAppendingFormat:@" %@", placemark.thoroughfare];
    }
  }
  
  if(placemark.areasOfInterest && placemark.areasOfInterest.count == 1) {
    simpleName = placemark.areasOfInterest.firstObject;
  }
  
  [self determineDefaultName:simpleName];
  NSString *address = ABCreateStringWithAddressDictionary(placemark.addressDictionary, YES);
  
  self.friendlyName = address;
}

@end

File: GeoNoter/Models/PQGPoint.swift
//
//  PQGPoint.swift
//  GeoNoter
//
//  Created by Patrick Quinn-Graham on 1/08/2014.
//  Copyright (c) 2014 Patrick Quinn-Graham. All rights reserved.
//

import Foundation
import MapKit
import AddressBookUI
import CoreLocation
import GeoNoterCore

final class PQGPoint: PQGModel {
  
  override class func tableName() -> String {
    return "point"
  }
  
  //MARK: - Private Variables
  private var _pointId:      Int64?
  private var _friendlyName: String?
  private var _name:         String?
  private var _memo:         String?
  private var _recordedAt:   Date?
  private var _latitude:     CLLocationDegrees?
  private var _longitude:    CLLocationDegrees?
  private var _foursquareId: String?
  
  //MARK: - Public properties
  
  var friendlyName : String? {
    get {
      return hydrate()._friendlyName
    }
    set (newFriendlyName) {
      hydrate()._friendlyName = newFriendlyName
    }
  }
  
  var name : String? {
    get {
      return hydrate()._name
    }
    set (name) {
      hydrate()._name = name
    }
  }
  
  var memo : String? {
    get {
      return hydrate()._memo
    }
    set (memo) {
      hydrate()._memo = memo
    }
  }
  
  var recordedAt : Date? {
    get {
      return hydrate()._recordedAt
    }
    set (recordedAt) {
      hydrate()._recordedAt = recordedAt
    }
  }
  
  var latitude : CLLocationDegrees? {
    get {
      return hydrate()._latitude
    }
    set (latitude) {
      hydrate()._latitude = latitude
    }
  }
  
  var longitude : CLLocationDegrees? {
    get {
      return hydrate()._longitude
    }
    set (longitude) {
      hydrate()._longitude = longitude
    }
  }
  
  var location: CLLocation? {
    get {
      if let lat = self.latitude {
        if let lng = self.longitude {
          return CLLocation(latitude: lat, longitude: lng)
        }
      }
      return nil
    }
    set (location) {
      if let coordinate = location?.coordinate {
        self.longitude = coordinate.longitude
        self.latitude = coordinate.latitude
      } else {
        self.longitude = nil
        self.latitude = nil
      }
    }
  }
  
  var foursquareId : String? {
    get {
      return hydrate()._foursquareId
    }
    set (foursquareId) {
      hydrate()._foursquareId = foursquareId
    }
  }
  
  var serializedForWatch: [NSString: AnyObject] {
    get {
      _ = hydrate()
      let id = NSNumber(value: primaryKey)
      
      var serialized = [
        "id": id,
        "name": name ?? "(no name)",
        "friendlyName": friendlyName ?? "",
        "memo": memo ?? ""
      ]
      
      if let location = location {
        serialized["location"] = [
          "lat": location.coordinate.latitude,
          "lng": location.coordinate.longitude
        ]
      }
      
      return serialized
    }
  }
  
  //MARK: - Lifecycle
  
  override func dehydrate() {
    _pointId = nil
    _friendlyName = nil
    _name = nil
    _memo = nil
    _recordedAt = nil
    _longitude = nil
    _foursquareId = nil
  }
  
  override func hydrateRequired(_ row: FMResultSet) {
    _friendlyName = row.string(forColumn: "friendly_name")
    _name         = row.string(forColumn: "name")
    _memo         = row.string(forColumn: "memo")
    _recordedAt   = row.date(forColumn: "recorded_at")
    _latitude     = row.double(forColumn: "latitude")
    _longitude    = row.double(forColumn: "longitude")
    _foursquareId = row.string(forColumn: "foursqure_venue_id")
  }
  
  override func saveForNew(_ db: FMDatabase) {
    if recordedAt == nil {
      recordedAt = Date()
    }
    let lat = numberWithDouble(latitude) ?? NSNull()
    let lon = numberWithDouble(longitude) ?? NSNull()    
    NSLog("saveForNew recordedAt \(recordedAt)")
    db.executeUpdate("INSERT INTO \(tableName())  (id, friendly_name, name, memo, recorded_at, latitude, longitude, foursqure_venue_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
      NSNumber(value: primaryKey),
      friendlyName ?? NSNull(),
      name ?? NSNull(),
      memo ?? NSNull(),
      recordedAt ?? NSNull(),
      lat,
      lon,
      foursquareId ?? NSNull()
    )
  }
  
  override func saveForUpdate(_ db: FMDatabase)  {
    let lat = numberWithDouble(latitude) ?? NSNull()
    let lon = numberWithDouble(longitude) ?? NSNull()
    db.executeUpdate("UPDATE \(tableName()) SET friendly_name = ?, name = ?, memo = ?, " +
      "recorded_at = ?, latitude = ?, longitude = ?, foursqure_venue_id = ? WHERE id = ?",
      friendlyName ?? NSNull(),
      name ?? NSNull(),
      memo ?? NSNull(),
      recordedAt ?? NSNull(),
      lat,
      lon,
      foursquareId ?? NSNull(),
      NSNumber(value: primaryKey)
    )
  }
  
  override func wasDestroyed() {
    store.points.removeCachedObject(primaryKey)
    store.withDatabase { db in
      db.executeUpdate("DELETE FROM point_tag WHERE tag_id = ?", NSNumber(value: self.primaryKey))

      return
    }
    
    for attachment in store.attachments.find.with("point_id = \(primaryKey)").all {
      attachment.destroy()
    }
  }
  
  //MARK: - Tags
  
  var tags : [PQGTag] {
  get {
    let cond = "id IN (select tag_id FROM point_tag WHERE point_id = \(primaryKey))"
    return store.tags.find.with(cond).all
  }
  set (tags) {
    removeAllTags()
    for tag in tags {
      addTag(tag)
    }
  }
  }
  
  func addTag(_ tag: PQGTag) {
    store.withDatabase { db in
      db.executeUpdate("INSERT INTO point_tag (point_id, tag_id) VALUES (?, ?)",
        NSNumber(value: self.primaryKey),
        NSNumber(value: tag.primaryKey)
      )
      return
    }
  }
  
  func removeTag(_ tag: PQGTag) {
    store.withDatabase { db in
      db.executeUpdate("DELETE FROM point_tag WHERE point_id = ? AND tag_id = ?",
        NSNumber(value: self.primaryKey),
        NSNumber(value: tag.primaryKey)
      )
      return
    }
  }
  
  private func removeAllTags() {
    store.withDatabase { db in
      db.executeUpdate("DELETE FROM point_tag WHERE point_id = ?",
        NSNumber(value: self.primaryKey)
      )
      return
    }
  }
  
  //MARK: - Attachments
  
  var attachments : [PQGAttachment] {
    return store.attachments.find.with("point_id = \(primaryKey)").all
  }
  
  func addAttachment(_ data: Data, withExtension fileExtension: String) -> PQGAttachment {
    let fileName = UUID().uuidString.appendingFormat(".%@", fileExtension)
    
    let actualFile = try! PQGPersistStore.attachmentsDirectory().appendingPathComponent(fileName)
    
    try? data.write(to: actualFile, options: [.dataWritingAtomic])
    
    let attachment = PQGAttachment(store: store)
    attachment.fileName = fileName
    attachment.kind = fileExtension
    attachment.point = self
    
    let dateFormatter = DateFormatter()
    dateFormatter.dateStyle = .mediumStyle
    dateFormatter.timeStyle = .mediumStyle
    
    let today = Date()
    
    attachment.friendlyName = "Picture - \(dateFormatter.string(from: today))"
    attachment.memo = "No memo"
    attachment.recordedAt = today
    attachment.save()
    
    return attachment
  }
  
  //MARK: - Point Helpers
  
  func determineDefaultName(_ locationName: String?) {
    let defaultName = UserDefaults.standard().string(forKey: "LocationsDefaultName")
    if defaultName == "most-specific" && locationName != nil {
      name = locationName
    } else if defaultName == "coords" {
      name = "\(longitude) \(latitude)"
      NSLog("Use lat/long name... \(name)");
    } else {
      let dateFormatter = DateFormatter()
      dateFormatter.dateStyle = .mediumStyle
      dateFormatter.timeStyle = .mediumStyle
      name = dateFormatter.string(from: Date())
      NSLog("Use date/time name... \(name)");
    }
  }
  
  func setupFromFoursquareVenue(_ venue: NSDictionary) {
    
    if let name = venue["name"] as? String {
      self.name = name
    } else {
      // TODO: Error
      return
    }
    
    if let location = venue["location"] as? NSDictionary {
      
      if let longitude = location["lng"] as? NSNumber {
        self.longitude = longitude.doubleValue
      }
      
      if let latitude = location["lat"] as? NSNumber {
        self.latitude = latitude.doubleValue
      }
      
      if let address = location["formattedAddress"] as? NSArray {
        var addressString = ""
        for line in address {
          if let addressLine = line as? String {
            addressString = addressString + "\n" + addressLine
          }
        }
        self.friendlyName = addressString
      }
      
    } else {
      // TODO: Error
      return
    }
    
    if let memo = venue["url"] as? String {
      self.memo = memo
    } else {
      self.memo = "No memo"
    }
    
    self.save()
  }
  
  func setupAsNewItem(_ coordinate: CLLocationCoordinate2D, completionHandler: (NSError?)->()) {
    NSLog("Long! \(coordinate.latitude)")
    NSLog("Lat! \(coordinate.longitude)")
    self.longitude = coordinate.longitude
    self.latitude = coordinate.latitude
    self.name = "Untitled"
    self.friendlyName = "Untitled"
    self.memo = "No memo"
    
    if UserDefaults.standard().bool(forKey: "LocationsUseGeocoder") {
      NSLog("Geocoder time")
      self.geocode { error in
        self.save()
        completionHandler(nil)
      }
    } else {
      NSLog("No geocoder here baby!")
      self.determineDefaultName(nil)
      self.save()
      completionHandler(nil)
    }
  }
  
  var bgTask : UIBackgroundTaskIdentifier = UIBackgroundTaskInvalid
  
  private func geocode(_ completionHandler: (NSError?)->()) {
    bgTask = UIApplication.shared().beginBackgroundTask {
      if self.bgTask != UIBackgroundTaskInvalid {
        UIApplication.shared().endBackgroundTask(self.bgTask)
        self.bgTask = UIBackgroundTaskInvalid
      }
    }
    
    let location = CLLocation(latitude: latitude!, longitude: longitude!)
    NSLog("Geocoding \(location)")
    
    LocationHelper.sharedHelper().geocode(location) { placemarks, error in
      UIApplication.shared().endBackgroundTask(self.bgTask)
      self.bgTask = UIBackgroundTaskInvalid
      if error != nil || placemarks!.count == 0 {
        self.friendlyName = "Geocoder Unavailable"
      } else {
        self.reverseGeocoderDidFindPlacemark(placemarks![0])
      }
      completionHandler(error)
    }
    
  }
  
  private func reverseGeocoderDidFindPlacemark(_ placemark: CLPlacemark) {
    var simpleName = ""
   
     func setIf(_ x: String?) -> Bool {
      if let unwrapped = x {
        if unwrapped != "" {
          simpleName = unwrapped
          return true
        }
      }
      return false
    }
    
    setIf(placemark.country)
    setIf(placemark.administrativeArea)
    setIf(placemark.locality)
    setIf(placemark.subLocality)
    
    if setIf(placemark.thoroughfare) {
      if setIf(placemark.subThoroughfare) {
        simpleName = placemark.subThoroughfare! + " " + placemark.thoroughfare!
      }
    }

    if placemark.areasOfInterest != nil && placemark.areasOfInterest?.count == 1 {
      setIf(placemark.areasOfInterest?[0])
    }

    determineDefaultName(simpleName)

    friendlyName = ABCreateStringWithAddressDictionary(placemark.addressDictionary!, true)

  }
  
}
